File: frontend/src/services/export.js
import api from './api'

/**
 * 文档导出服务
 * 纯前端实现 Markdown 导出
 * Word/PDF/LaTeX 需要后端支持（已预留接口）
 */

export class ExportService {
  /**
   * TipTap JSON 转 Markdown（纯前端）
   */
  static toMarkdown(json) {
    if (!json || !json.content) return ''
    return this.processNodes(json.content)
  }

  static processNodes(nodes, listLevel = 0) {
    if (!nodes) return ''
    return nodes.map(node => this.processNode(node, listLevel)).join('')
  }

  static processNode(node, listLevel = 0) {
    switch (node.type) {
      case 'heading':
        const level = node.attrs?.level || 1
        const prefix = '#'.repeat(level)
        return `${prefix} ${this.processInline(node.content)}\n\n`

      case 'paragraph':
        const text = this.processInline(node.content)
        return text ? `${text}\n\n` : '\n'

      case 'bulletList':
        return this.processList(node.content, false, listLevel) + '\n'

      case 'orderedList':
        return this.processList(node.content, true, listLevel) + '\n'

      case 'listItem':
        return this.processNodes(node.content, listLevel)

      case 'blockquote':
        const quoteContent = this.processNodes(node.content)
        return quoteContent.split('\n').map(line => line ? `> ${line}` : '>').join('\n') + '\n'

      case 'codeBlock':
        const lang = node.attrs?.language || ''
        const code = this.processInline(node.content)
        return `\`\`\`${lang}\n${code}\n\`\`\`\n\n`

      case 'horizontalRule':
        return '---\n\n'

      case 'image':
        const src = node.attrs?.src || ''
        const alt = node.attrs?.alt || ''
        return `![${alt}](${src})\n\n`

      case 'table':
        return this.processTable(node) + '\n'

      default:
        if (node.content) {
          return this.processNodes(node.content, listLevel)
        }
        return ''
    }
  }

  static processList(items, ordered, level) {
    const indent = '  '.repeat(level)
    return items.map((item, index) => {
      const prefix = ordered ? `${index + 1}.` : '-'
      const content = this.processNodes(item.content, level + 1).trim()
      return `${indent}${prefix} ${content}`
    }).join('\n')
  }

  static processInline(nodes) {
    if (!nodes) return ''
    return nodes.map(node => {
      let text = node.text || ''
      
      if (node.marks) {
        node.marks.forEach(mark => {
          switch (mark.type) {
            case 'bold':
              text = `**${text}**`
              break
            case 'italic':
              text = `*${text}*`
              break
            case 'underline':
              text = `<u>${text}</u>`
              break
            case 'code':
              text = `\`${text}\``
              break
            case 'link':
              text = `[${text}](${mark.attrs?.href || ''})`
              break
          }
        })
      }
      
      return text
    }).join('')
  }

  static processTable(tableNode) {
    if (!tableNode.content) return ''
    
    const rows = tableNode.content
    let markdown = ''
    
    rows.forEach((row, rowIndex) => {
      const cells = row.content || []
      const cellContents = cells.map(cell => {
        return this.processInline(cell.content?.[0]?.content || []).trim()
      })
      
      markdown += '| ' + cellContents.join(' | ') + ' |\n'
      
      // 添加表头分隔符
      if (rowIndex === 0) {
        markdown += '| ' + cellContents.map(() => '---').join(' | ') + ' |\n'
      }
    })
    
    return markdown
  }

  /**
   * 下载 Markdown 文件
   */
  static downloadMarkdown(json, filename = 'document.md') {
    const markdown = this.toMarkdown(json)
    const blob = new Blob([markdown], { type: 'text/markdown;charset=utf-8' })
    const url = URL.createObjectURL(blob)
    
    const a = document.createElement('a')
    a.href = url
    a.download = filename
    document.body.appendChild(a)
    a.click()
    document.body.removeChild(a)
    URL.revokeObjectURL(url)
    
    return true
  }

  /**
   * 导出为 Word（调用后端）
   */
  static async toWord(json, filename = 'document.docx') {
    try {
      const response = await api.post('/export/docx', {
        content: json,
        title: filename.replace('.docx', '')
      }, {
        responseType: 'blob'
      })

      const blob = new Blob([response.data], { type: 'application/vnd.openxmlformats-officedocument.wordprocessingml.document' })
      const url = window.URL.createObjectURL(blob)
      
      const a = document.createElement('a')
      a.href = url
      a.download = filename
      document.body.appendChild(a)
      a.click()
      document.body.removeChild(a)
      window.URL.revokeObjectURL(url)
      
      return true
    } catch (error) {
           console.error('Export docx error', error);
      throw new Error('Word 导出失败: ' + (error.response?.data?.error || error.message))
    }
  }


  static downloadWord(fileBlob, filename = 'document.docx') {
  }

  /**
   * 导出为 PDF（需要后端）
   * @placeholder 预留接口
   */
  static async toPDF(json, wsService) {
    // TODO: 后端实现
    throw new Error('PDF 导出需要后端支持')
  }

  /**
   * 导出为 LaTeX（需要后端）
   * @placeholder 预留接口
   */
  static async toLaTeX(json, wsService) {
    // TODO: 后端实现
    throw new Error('LaTeX 导出需要后端支持')
  }
}

File: backend/routes/export.py
from flask import Blueprint, request, send_file, jsonify
from docx import Document
import io

export_bp = Blueprint('export', __name__, url_prefix='/api/export')


# ==============================
# docx 导出相关函数
# ==============================

def process_node(doc, node):
    """递归处理 TipTap JSON 节点并写入 docx"""
    node_type = node.get('type')
    content = node.get('content', [])
    
    # 提取纯文本内容
    text = ""
    
    # 根据节点类型处理
    if node_type == 'doc':
        for child in content:
            process_node(doc, child)
            
    elif node_type == 'paragraph':
        # 如果是空的段落，也需要添加
        p = doc.add_paragraph()
        for child in content:
            if child.get('type') == 'text':
                run = p.add_run(child.get('text', ''))
                # 处理加粗、斜体等 marks
                if 'marks' in child:
                    for mark in child['marks']:
                        if mark['type'] == 'bold': run.bold = True
                        if mark['type'] == 'italic': run.italic = True
                        # 可以添加更多样式支持，如下划线等
                        if mark['type'] == 'underline': run.underline = True
                        
    elif node_type == 'heading':
        level = node.get('attrs', {}).get('level', 1)
        # 提取标题文本
        heading_text = ""
        for child in content:
            if child.get('type') == 'text':
                heading_text += child.get('text', '')
        
        doc.add_heading(heading_text, level=level)
        
    elif node_type == 'bulletList':
        for child in content:
            if child.get('type') == 'listItem':
                # docx 库对列表支持比较简单，通常只能通过 paragraph style 控制
                # 这里简化处理，直接用 List Bullet 样式
                # 注意：嵌套列表在这个简化版可能无法完美展现
                p = doc.add_paragraph(style='List Bullet')
                # 列表项里面通常是 paragraph
                for item_content in child.get('content', []):
                    if item_content.get('type') == 'paragraph':
                        for text_node in item_content.get('content', []):
                            if text_node.get('type') == 'text':
                                run = p.add_run(text_node.get('text', ''))
                                if 'marks' in text_node:
                                    for mark in text_node['marks']:
                                        if mark['type'] == 'bold': run.bold = True
                                        if mark['type'] == 'italic': run.italic = True
                                        
    elif node_type == 'orderedList':
        for child in content:
            if child.get('type') == 'listItem':
                p = doc.add_paragraph(style='List Number')
                for item_content in child.get('content', []):
                    if item_content.get('type') == 'paragraph':
                        for text_node in item_content.get('content', []):
                            if text_node.get('type') == 'text':
                                run = p.add_run(text_node.get('text', ''))
                                if 'marks' in text_node:
                                    for mark in text_node['marks']:
                                        if mark['type'] == 'bold': run.bold = True
                                        if mark['type'] == 'italic': run.italic = True

    elif node_type == 'codeBlock':
       # 代码块用特定样式或单倍行距+固定字体
       # 暂时简化为普通文本
        code_text = ""
        for child in content:
            if child.get('type') == 'text':
               code_text += child.get('text', '')
        
        if code_text:
            p = doc.add_paragraph(code_text)
            p.style = 'No Spacing' # 紧凑样式

    elif node_type == 'blockquote':
        # 引用块
         for child in content:
            if child.get('type') == 'paragraph':
                text = ""
                for text_node in child.get('content', []):
                    text += text_node.get('text', '')
                if text:
                    p = doc.add_paragraph(text, style='Quote')


@export_bp.route('/docx', methods=['POST'])
def export_docx():
    try:
        data = request.json
        if not data:
            return jsonify({'error': 'No data provided'}), 400
            
        title = data.get('title', 'Document')
        editor_json = data.get('content', {})

        # 创建 Word 文档
        doc = Document()
        doc.add_heading(title, 0)

        # 递归处理内容
        if not editor_json or not isinstance(editor_json, dict):
             process_node(doc, {'type': 'doc', 'content': []}) # TODO_202 应该返回报错
        else:
             process_node(doc, editor_json)

        # 保存到内存流
        file_stream = io.BytesIO()
        doc.save(file_stream)
        file_stream.seek(0)

        return send_file(
            file_stream,
            as_attachment=True,
            download_name=f"{title}.docx",
            mimetype='application/vnd.openxmlformats-officedocument.wordprocessingml.document'
        )

    except Exception as e:
        print(f"Export error: {e}")
        return jsonify({'error': str(e)}), 500
